core/chains/evm/logpoller/parser.go

```diff
@@ -0,0 +1,499 @@
+	if err != nil {
+		return "", err
+	}
+
+	return fmt.Sprintf(
+		pattern,
+		subfield,
+		cmp,
+		args.withIndexedField(field, common.HexToHash(comp.Value)),
+	), nil
+}
+
+func (v *pgDSLParser) buildQuery(chainID *big.Int, expressions []query.Expression, limiter query.LimitAndSort) (string, *queryArgs, error) {
+	// reset transient properties
+	v.args = newQueryArgs(chainID)
+	v.expression = ""
+	v.err = nil
+
+	// build the query string
+	clauses := []string{"SELECT evm.logs.* FROM evm.logs"}
+
+	where, err := v.whereClause(expressions, limiter)
+	if err != nil {
+		return "", nil, err
+	}
+
+	clauses = append(clauses, where)
+
+	order, err := v.orderClause(limiter)
+	if err != nil {
+		return "", nil, err
+	}
+
+	if len(order) > 0 {
+		clauses = append(clauses, order)
+	}
+
+	limit := v.limitClause(limiter)
+	if len(limit) > 0 {
+		clauses = append(clauses, limit)
+	}
+
+	return strings.Join(clauses, " "), v.args, nil
+}
+
+func (v *pgDSLParser) whereClause(expressions []query.Expression, limiter query.LimitAndSort) (string, error) {
+	segment := "WHERE evm_chain_id = :evm_chain_id"
+
+	if len(expressions) > 0 {
+		exp, err := v.combineExpressions(expressions, query.AND)
+		if err != nil {
+			return "", err
+		}
+
+		segment = fmt.Sprintf("%s AND %s", segment, exp)
+	}
+
+	if limiter.HasCursorLimit() {
+		var op string
+		switch limiter.Limit.CursorDirection {
+		case query.CursorFollowing:
+			op = ">"
+		case query.CursorPrevious:
+			op = "<"
+		default:
+			return "", errors.New("invalid cursor direction")
+		}
+
+		block, txHash, logIdx, err := valuesFromCursor(limiter.Limit.Cursor)
+		if err != nil {
+			return "", err
+		}
+
+		segment = fmt.Sprintf("%s AND block_number %s= :cursor_block AND tx_hash %s= :cursor_txhash AND log_index %s :cursor_log_index", segment, op, op, op)
+
+		v.args.withField("cursor_block_number", block).
+			withField("cursor_txhash", common.HexToHash(txHash)).
+			withField("cursor_log_index", logIdx)
+	}
+
+	return segment, nil
+}
+
+func (v *pgDSLParser) orderClause(limiter query.LimitAndSort) (string, error) {
+	sorting := limiter.SortBy
+
+	if limiter.HasCursorLimit() && !limiter.HasSequenceSort() {
+		var dir query.SortDirection
+
+		switch limiter.Limit.CursorDirection {
+		case query.CursorFollowing:
+			dir = query.Asc
+		case query.CursorPrevious:
+			dir = query.Desc
+		default:
+			return "", errors.New("unexpected cursor direction")
+		}
+
+		sorting = append(sorting, query.NewSortBySequence(dir))
+	}
+
+	if len(sorting) == 0 {
+		return "", nil
+	}
+
+	sort := make([]string, len(sorting))
+
+	for idx, sorted := range sorting {
+		var name string
+
+		order, err := orderToString(sorted.GetDirection())
+		if err != nil {
+			return "", err
+		}
+
+		switch sorted.(type) {
+		case query.SortByBlock:
+			name = blockFieldName
+		case query.SortBySequence:
+			sort[idx] = fmt.Sprintf("block_number %s, tx_hash %s, log_index %s", order, order, order)
+
+			continue
+		case query.SortByTimestamp:
```

Comment:
It might be a bit yagni, because we never sort by timestamp, right?

Reply:
Not in the current LP, but because it exists in the `QueryKey` definition, it is implemented here.